frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/toolbar/DiscoverAction.java

```diff
@@ -81,40 +84,58 @@ public void actionPerformed(AnActionEvent anActionEvent) {
             /* hasBranchActionToolTips */ false).show(), NON_MODAL));
   }
 
+  @UIEffect
   private static void openMacheteFile(Project project, GitRepository gitRepository) {
-    GuiUtils.invokeLaterIfNeeded(() -> {
-      var file = GitVfsUtils.getMacheteFile(gitRepository);
-      if (file.isDefined()) {
-        OpenFileAction.openFile(file.get(), project);
-      } else {
-        VcsNotifier.getInstance(project)
-            .notifyError(
-                /* title */ getString("action.GitMachete.OpenMacheteFileAction.notification.title.machete-file-not-found"),
-                /* message */ format(
-                    getString("action.GitMachete.OpenMacheteFileAction.notification.message.machete-file-not-found"),
-                    gitRepository.getRoot().getPath()));
-      }
-    }, NON_MODAL);
+    var file = GitVfsUtils.getMacheteFile(gitRepository);
+    if (file.isDefined()) {
+      OpenFileAction.openFile(file.get(), project);
+    } else {
+      VcsNotifier.getInstance(project)
+          .notifyError(
+              /* title */ getString("action.GitMachete.OpenMacheteFileAction.notification.title.machete-file-not-found"),
+              /* message */ format(
+                  getString("action.GitMachete.OpenMacheteFileAction.notification.message.machete-file-not-found"),
+                  gitRepository.getRoot().getPath()));
+    }
   }
 
   private void saveDiscoveredLayout(IGitMacheteRepositorySnapshot repositorySnapshot, Path macheteFilePath, Project project,
-      BaseEnhancedGraphTable baseEnhancedGraphTable, IBranchLayoutWriter branchLayoutWriter) {
+      BaseEnhancedGraphTable baseEnhancedGraphTable, IBranchLayoutWriter branchLayoutWriter, @UI Runnable postWriteRunnable) {
     var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
     if (branchLayout == null) {
       VcsNotifier.getInstance(project).notifyError(
           /* title */ getString("action.GitMachete.DiscoverAction.notification.title.cannot-discover-layout-error"),
           /* message */ "");
       return;
     }
+
     new Task.Backgroundable(project, getString("action.GitMachete.DiscoverAction.write-file.task-title")) {
       @Override
       public void run(ProgressIndicator indicator) {
-        Try.run(() -> branchLayoutWriter.write(macheteFilePath, branchLayout, /* backupOldLayout */ true))
-            .onFailure(e -> VcsNotifier.getInstance(project).notifyError(
-                /* title */ getString("action.GitMachete.DiscoverAction.notification.title.write-file-error"),
-                /* message */ e.getMessage() != null ? e.getMessage() : ""))
-            .onSuccess(__ -> baseEnhancedGraphTable.queueRepositoryUpdateAndModelRefresh());
+        try {
+          branchLayoutWriter.write(macheteFilePath, branchLayout, /* backupOldLayout */ true);
+        } catch (Exception e) {
+          throw new RuntimeException(e);
```

Comment:
Or just `@SneakyThrows` ;) actually... it's weird that `Task#run` signature does NOT include any `throws` clause... given that this API explicitly accounts for failures via `onThrowable`

Reply:
Just another quirk of checked exceptions in general ;)

Reply:
amended

Reply:
Ok, approving the PR in general, lemme just test it locally once CI passes